### DISCUSSION TITLE
How to Change Application Language at Runtime?

### DISCUSSION THREAD

**Discussion by ghost1372** [microsoft/WindowsAppSDK#1404] [Q&A] [answered]
I want to change application language at runtime, i can change language when application is startup like this:
```
protected override void OnLaunched(Microsoft.UI.Xaml.LaunchActivatedEventArgs args)
{
    ApplicationLanguages.PrimaryLanguageOverride = "en-US";
    m_window = new MainWindow();
    m_window.Activate();
}
```

For UWP, the following code seems to do this
```
ApplicationLanguages.PrimaryLanguageOverride = "fa-IR";
Windows.ApplicationModel.Resources.Core.ResourceContext.GetForCurrentView().Reset();
Windows.ApplicationModel.Resources.Core.ResourceContext.GetForViewIndependentUse().Reset();
Frame.Navigate(this.GetType());
```

but for WinUI 3 There are problems
if we write this code in **MainWindow.xaml.cs** :
Frame.**Navigate** is not exist

if we move this code to a Page, **Frame.Navigate** is exist but is **null**

also In any case, **GetForCurrentView** get exception `System.Runtime.InteropServices.COMException: 'Resource Contexts may not be created on threads that do not have a CoreWindow. (0x80073B27)'`

Can anyone help and rewrite this code for WinUI 3?

**Comment by axelandrejs** [comment, marked as answer]
To resolve resources in a different language, you need to set the resource context using the new WinAppSdk APIs. You can find a sample here: https://github.com/microsoft/WindowsAppSDK-Samples/blob/main/Samples/ResourceManagement/cs-winui/winui_desktop_packaged_app/MainWindow.xaml.cs

**Reply by oliverw** [reply]
> To resolve resources in a different language, you need to set the resource context using the new WinAppSdk APIs. You can find a sample here: https://github.com/microsoft/WindowsAppSDK-Samples/blob/main/Samples/ResourceManagement/cs-winui/winui_desktop_packaged_app/MainWindow.xaml.cs

This does not help at all as this only affects resources resolved directly through the manually created manager/loader. Then there is `ApplicationLanguages.PrimaryLanguageOverride` which also does not work reliably as it currently takes two application restarts to have the changed value apply. What a shitshow.
